Ticket31240 merged 1

nmathewson opened the pull request.

nmathewson committed:
Add a config_mgr_t type to wrap config_format_t

Remember that our goal in the present refactoring is to allow each
subsystem to declare its own configuration structure and
variables.  To do this, each module will get its own
config_format_t, and so we'll want a different structure that wraps
several config_format_t objects.  This is a "config_mgr_t".

nmathewson committed:
Start teaching config_mgr_t to handle sub-objects and sub-formats

The eventual design here will be that multiple config_format_t
objects get registered with a single config_mgr_t.  That
config_mgr_t manages a "top-level" object, which has a pointer to
the other objects.

I had earlier thought of a different design, where there would be no
top-level object, and config_mgr_t would deal with a container
instead.  But this would require a bunch of invasive refactoring
that I don't think we should do just yet.

nmathewson committed:
Replace config_find_option with a variant that exposes less

nmathewson committed:
Lower responsibility for listing changed options into confparse.c

nmathewson committed:
Fix every place in config.c that knew about option_vars_.

Iterating over this array was once a good idea, but now that we are
going to have a separate structure for each submodule's
configuration variables, we should indirect through the config_mgr_t
object.

nmathewson committed:
Change CONFIG_CHECK() macro to not need a config_format_t

We'll want it to check all the subsidiary structures of the
options object.

nmathewson committed:
Add a "freeze" function for config_mgr_t objects.

It's important to make sure that we don't change a config_mgr_t
after we start using it to make objects, or we could get into
inconsistent states.  This feature is the start of a safety
mechanism to prevent this problem.

nmathewson committed:
Sort variables in config_mgr_t.all_vars alphabetically

nmathewson committed:
Use config_new() to construct configuration objects.

We'll need to do it this way once the objects become more complex.

nmathewson committed:
Use special magic to enforce manager/object connection.

Every time we finalize a config manager, we now generate a new magic
number for it, so that we'll get an assertion failure if we ever two
to use an object with a different configuration manager than the one
that generated it.

nmathewson committed:
Make a config_suite_t type to hold multiple config sub-objects

Right now, it doesn't do anything; this patch is meant to make sure
that we're doing memory management correctly.

nmathewson committed:
Refactor config free logic to use a single path.

The right way to free a config object is now to wrap config_free(),
always.  Instead of creating an alternative free function, objects
should provide an alternative clear callback to free any fields that
the configuration manager doesn't manage.

This lets us simplify our code a little, and lets us extend the
confparse.c code to manage additional fields in config_free.

nmathewson committed:
Partial support for multiplicity in configuration objects

A configuration manager, in addition to a top-level format object,
may now also know about a suite of sub-formats.  Top-level
configuration objects, in turn, may now have a suite of
sub-objects.

nmathewson committed:
Merge branch 'ticket31240' into ticket31240_merged_1

teor2345 commented on src/app/config/config.c:
```diff
@@ -977,14 +990,14 @@ set_options(or_options_t *new_val, char **msg)
     elements = smartlist_new();
     for (i=0; options_format.vars[i].member.name; ++i) {
-      const config_var_t *var = &options_format.vars[i];
+      const config_var_t *var = &options_format.vars[i]; // XXXX 29211
```
What does this comment mean?

I don't know if this is a blocking issue, or which ticket it blocks.

nmathewson replied:
Sorry, that means "fix this before 29211 is done."

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -271,14 +310,14 @@ config_assign_line(const config_format_t *fmt, void *options,
     }
     return 0;
   } else if (c->command == CONFIG_LINE_CLEAR && !clear_first) {
     // XXXX This is unreachable, since a CLEAR line always has an
```
Why are the XXXXs here?
Does something still need to be done?

I think this is cleanup that should be done before closing #29211.

nmathewson replied:
Agreed. Opened https://trac.torproject.org/projects/tor/ticket/31529#ticket

teor2345 commented on src/app/config/confparse.h:
```diff
@@ -72,34 +87,34 @@ typedef struct config_format_t {
-    config_free_((fmt), (options));                   \
+void *config_new(const config_mgr_t *fmt);
+void config_free_(const config_mgr_t *fmt, void *options);
+#define config_free(mgr, options) do {                \
```
Should this macro uses STMT_BEGIN and STMT_END?

This is cleanup that should be done before closing #29211.

nmathewson replied:
We have over 175 other "} while (0)" occurrences in our code, and only 146 STMT_ENDs.  I've opened https://trac.torproject.org/projects/tor/ticket/31530 for this.

teor2345 commented on src/app/config/config.c:
```diff
@@ -843,15 +843,15 @@ static void config_maybe_load_geoip_files_(const or_options_t *options,
 static int options_validate_cb(void *old_options, void *options,
```
Commit comment:

In future, please split mass change commits into at least two commits:
* automated changes using sed or coccinelle scripts
* manual cleanup changes

This does not block reviews for code that is already in commits or branches. But it should block reviews for any new commits.

teor2345 replied:
Opened trac #31511 for this task, please resolve this comment once you've read it.

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -37,8 +37,38 @@
+static void
+managed_var_free_(managed_var_t *mv)
+{
+  if (!mv)
```
This if statement is unnecessary: tor_free(NULL) does nothing.

This change should be made before #29211 is closed.

nmathewson replied:
I prefer to have this check in every free function, so that if we add more fields, we will not forget to add the check.  The compiler will (I think) notice that it is redundant, and will not add any redundant code.

teor2345 replied:
Ok

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -48,24 +78,99 @@ struct config_mgr_t {
+  /** A smartlist of managed_var_t objects for all configuration formats. */
+  smartlist_t *all_vars;
+  /** A smartlist of config_abbrev_t objects for all abbreviations. These
+   * objects are */
```
Please complete this comment.

This change should be made before #29211 is closed.

nmathewson replied:
Added a fix in  9fe9f20a57253e

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -48,24 +78,99 @@ struct config_mgr_t {
+ * NOTE: XXXX Eventually, there will be multiple objects supported within the
+ * toplevel object.  For example, the or_options_t will contain pointers
+ * to configuration objects for other modules.  This function gets
+ * the sub-object for a particular modules.
```
Typo: module

This change should be made before #29211 is closed.

teor2345 replied:
Did you miss this comment?

nmathewson replied:
I did! Fixed in 0eb81ec26358c9

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -289,44 +416,47 @@ config_assign_line(const config_mgr_t *mgr, void *options,
     }
     return 0;
   } else if (c->command == CONFIG_LINE_CLEAR && !clear_first) {
     // XXXX This is unreachable, since a CLEAR line always has an
```
What needs to be done here?
Or should the XXXX be removed?

I don't know what this change should be, or what it blocks.

nmathewson replied:
This is also https://trac.torproject.org/projects/tor/ticket/31529#ticket

teor2345 reviewed (changes requested):
This is a partial review of the first few commits.

I have not reviewed "Replace config_find_option with..." or any of the commits after it.

teor2345 commented on src/app/config/config.c:
```diff
@@ -988,35 +985,16 @@ set_options(or_options_t *new_val, char **msg)
+      smartlist_add(elements, line->key);
+      smartlist_add(elements, line->value);
     }
     control_event_conf_changed(elements);
```
Wow, smartlist(k, v, k, v, …) is an unexpected API.
It's legacy code, so it's not something you need to fix.
But eventually maybe we could do smartlist(k, k, …), smartlist(v, v, …).

We can open a ticket to do this sometime in the future, if you think it is a good idea.

nmathewson replied:
I agree a better API would be good; a config_line_t would actually be a reasonable type to use here.

teor2345 replied:
I opened https://trac.torproject.org/projects/tor/ticket/31531

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -738,6 +738,46 @@ config_is_same(const config_mgr_t *mgr,
+ **/
+config_line_t *
+config_get_changes(const config_mgr_t *mgr,
+                   const void *options1, const void *options2)
```
Does this stop being void * later in the refactor?
If not, please document the possible types of options1 and options2.

This change should block #29211, but we can merge this PR without it.

nmathewson replied:
16f07f1acb05d2 has documentation here.

teor2345 commented on src/app/config/config.c:
```diff
@@ -8125,34 +8125,34 @@ getinfo_helper_config(control_connection_t *conn,
-        if (strcmp(option_vars_[i].member.name, "FallbackDir") == 0) {
+        if (strcmp(var->member.name, "FallbackDir") == 0) {
           /*
            * Similarly count fallback lines, so that we can decided later
```
Typo: decide

Typos don't block anything, but it would be nice to fix them in #29211.

nmathewson replied:
fb85c089325eb1 has a fix.

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -175,6 +175,31 @@ config_mgr_free_(config_mgr_t *mgr)
 
+/** Return a new smartlist_t containing a config_var_t for every variable that
+ * <b>mgr</b> knows about.  The elements of this smartlist do not need
+ * to be freed. */
```
But how long do they remain valid?

nmathewson replied:
addressed in 4d1dfabf2db98b

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -175,6 +175,31 @@ config_mgr_free_(config_mgr_t *mgr)
+}
+
+/** Return a new smartlist_t containing the names of all deprecated variables.
+ * The elements of this smartlist do not need to be freed. */
```
But how long do they remain valid?

nmathewson replied:
addressed in 4d1dfabf2db98b

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -85,6 +85,10 @@ struct config_mgr_t {
+  /** True if this manager has been frozen and cannot have any more formats
+   * added to it. A manager must be frozen before it can be used to construct
+   * or manipulate objects. */
+  bool frozen;
```
This could be a 1-bit flag, if we ever add any more flags.
But we would have to add 4-8 bools to even make a difference to the memory usage here.

nmathewson replied:
agreed.

teor2345 commented on src/feature/dirauth/shared_random_state.c:
```diff
@@ -599,11 +598,12 @@ disk_state_reset(void)
+  /* Clear other fields. */
+  sr_disk_state->ValidAfter = 0;
+  sr_disk_state->ValidUntil = 0;
+  sr_disk_state->Version = 0;
```
When we add fields, how will we make sure we remember to clear them all?

nmathewson replied:
We don't need to do so in general, the manager will take care of it for us.  This is only necessary for derived fields, and fields for which the unit tests are expecting an unusual behavior, IIRC.

(I can dig deeper here; I forget the exact reason why this function is clearing these, but I think it is because of legacy cruft.)

teor2345 replied:
Yeah let's double-check what's going on here.

teor2345 replied:
Opened a ticket for follow-up:
https://trac.torproject.org/projects/tor/ticket/31544

I think this issue might be fixed in #31241 ???

teor2345 replied:
@nmathewson is this issue fixed in #31241?
Or should I leave #31544 open for follow up?

nmathewson replied:
Let's leave #31544 open for follow-up: it'll be easy enough to close it if we turn out to fix it in #31241.  I expect that #31241 is going to get a large rewrite,  so it's a good idea to keep #31544 open as a reminder.

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -37,6 +37,8 @@
 #include "lib/string/printf.h"
```
Commit typo: two -> try

nmathewson replied:
I think this is no longer in the code?  I see no wrong two/try instances in the current confparse.c

teor2345 replied:
One of the commit messages has this typo.

Unfortunately comments aren't linked to commits on GitHub, so I'll have to search for it.

teor2345 replied:
commit 94bfbf7859
Author: Nick Mathewson <nickm@torproject.org>
Date:   Tue Jul 23 10:41:57 2019 -0400

    Use special magic to enforce manager/object connection.
    
    Every time we finalize a config manager, we now generate a new magic
    number for it, so that we'll get an assertion failure if we ever two
    to use an object with a different configuration manager than the one
    that generated it.

This commit message typo blocks merging this branch.

teor2345 replied:
@nmathewson I think you fixed this commit message in the revised branch, please confirm.

nmathewson replied:
yes.

teor2345 commented on src/test/test_confparse.c:
```diff
@@ -158,7 +158,6 @@ test_confparse_init(void *arg)
   config_init(mgr, tst);
 
   // Make sure that options are initialized right. */
-  tt_uint_op(tst->magic, OP_EQ, TEST_MAGIC);
```
Should there be a replacement for this test?
How do we know the values are unique?
(That is, how do the tests ensure that any code changes preserve the uniqueness of toplevel_magic?)

I think this change should block the merge of this PR: if we don't have uniqueness, there is something fundamentally wrong with this PR. (And if any future PR breaks uniqueness, we want to know, so that PR does not get merged.)

nmathewson replied:
To make sure I understand: I think you're asking for a test to make sure that our code for making toplevel_magic settings actually results in different toplevel_magic settings that are different across objects.

I think it will be okay if _usually_ they are unique, since our only use for them is to assert that they are what we expect them to be?

teor2345 replied:
Yes. But the probability of two values being the same is 1 in 2^32, so we should be able to use that in a test.

teor2345 replied:
@nmathewson what was the resolution for this issue?

If it is possible to add the runtime magic generation uniqueness tests now, they should block this branch.

teor2345 replied:
I think this is 55b7280ebd8d56854b069f32397b3a544e1f1c39?

nmathewson replied:
Yes.

teor2345 commented on src/app/config/confparse.c:
```diff
@@ -72,6 +72,36 @@ managed_var_free_(managed_var_t *mv)
+static void
+config_suite_free_(config_suite_t *suite)
+{
+  if (!suite)
```
This check is unnecessary, tor_free(NULL) does nothing.

This is not a blocking change, but we should clean it up before we merge #29211.

nmathewson replied:
As above, I prefer this for future-proofing, but I'm happy to discuss it if you disagree.

teor2345 replied:
I don't mind.

teor2345 commented on src/app/config/confparse.h:
```diff
@@ -61,6 +61,9 @@ typedef struct config_format_t {
   const struct_member_t *extra;
+  /** The position of a config_suite_t pointer within the toplevel object,
+   * or -1 if there is no such pointer. */
+  int config_suite_offset;
```
Shouldn't this be an offset_t or ptrdiff_t?
Is int guaranteed to be large enough to hold all pointer offsets?
Will some compilers complain that it's not?

nmathewson replied:
ptrdiff_t does seem better.

We already use int for this in struct_member_t, though, so I'm opening a ticket to convert them both.  https://trac.torproject.org/projects/tor/ticket/31532#ticket

teor2345 commented on src/app/config/or_options_st.h:
```diff
@@ -1107,6 +1108,9 @@ struct or_options_t {
    **/
   int DormantCanceledByStartup;
+
+  /**DOCDOC*/
```
Please document

I don't know if this is a blocking change yet. If we add actual data to subconfigs_, it should block this PR.

nmathewson replied:
ad22fbcfcd3f9e documents this.

teor2345 replied:
Ok, seems fine.

teor2345 commented on src/app/config/or_state_st.h:
```diff
@@ -94,6 +95,9 @@ struct or_state_t {
    * weren't.  "auto" on initial startup. */
   int Dormant;
+
+  /**DOCDOC*/
```
Please document.

nmathewson replied:
0c7528f96a8469 documents this.

teor2345 commented on src/test/test_confparse.c:
```diff
@@ -815,6 +816,7 @@ static config_format_t etest_fmt = {
   test_validate_cb,
   NULL,
   &extra,
+  -1,
```
There is no unit test with a subconfig: please add one.

This test should block this PR, because we want to ensure that the simplest subconfig works, and the memory management works.

teor2345 replied:
@nmathewson what was the resolution here?

teor2345 replied:
I think this is 55b7280ebd8d56854b069f32397b3a544e1f1c39, please confirm?

nmathewson replied:
That is correct.

teor2345 commented on src/app/config/confparse.h:
```diff
@@ -41,8 +41,8 @@ typedef int (*validate_fn_t)(void*,void*,void*,int,char**);
 
-/** Callback to free a configuration object. */
-typedef void (*free_cfg_fn_t)(const struct config_mgr_t *mgr, void*);
+/** Callback to clear all non-managed fields of a configuration object. */
```
Please explain what a non-managed field is.

This comment should not block this PR, but I need to understand what a non-managed field is.

nmathewson replied:
2a088ca392a5c5 tries to explain this.

teor2345 replied:
Ok. Do we need unit tests that have non-managed fields?

teor2345 replied:
@nmathewson do we need unit tests that have non-managed fields?

teor2345 replied:
This looks like beaf7305a0e18304a395faae5dd79bf1f94d1160?

nmathewson replied:
That is correct.

teor2345 commented on src/test/test_confparse.c:
```diff
@@ -824,7 +813,7 @@ static config_format_t etest_fmt = {
   test_vars,
   test_validate_cb,
-  test_free_cb,
+  NULL,
```
There aren't any tests that use a clear callback. Please add some.

teor2345 replied:
@nmathewson do we need unit tests that use a clear callback?

teor2345 replied:
This also looks like beaf7305a0e18304a395faae5dd79bf1f94d1160?

nmathewson replied:
That's right.

teor2345 reviewed (changes requested):
There are a bunch of missing unit tests for new callbacks.

The design looks pretty ok, but it's hard to tell without an overview of where we are going. (I would really like a diagram or a formatted text summary.)

nmathewson committed:
finish a comment explaining all_abbrevs

nmathewson committed:
Document types for config_get_changes()

nmathewson committed:
Fix typo in comment about FallbackDir handling

nmathewson committed:
document return lifespans for config_mgr_list_vars()

nmathewson committed:
Document or_options_t.subconfigs_

nmathewson committed:
Document or_state_t.substates_

nmathewson committed:
Better explain config_clear_fn_t

nmathewson committed:
Add tests for variable-listing functions.

This discovered a bug related to an extra & in
config_mgr_list_deprecated_vars(): fix that.

nmathewson committed:
Tests for config_find_option_name()

Fix a bug in config_find_option_name() where it did not consider
the abbreviations table.

nmathewson committed:
Start on test cases for the multi-object feature of confmgr.

This test case, at this point, only constructs the confmgr object.
More code to come.

nmathewson committed:
Confmgr tests: test deprecations and abbreviations.

nmathewson committed:
Simple tests for nested configuration formats

One test makes sure that the toplevel magic numbers are distinct.

One test makes sure that we can parse a configuration object
with two sub-objects.

nmathewson committed:
Add a test for dumping a compound configuration

nmathewson committed:
Fix typo in a comment on config_mgr_get_obj_mutable()

nmathewson committed:
Add a test for clear_fn support in handling non-managed fields

teor2345 commented on src/app/config/confparse.h:
```diff
@@ -41,7 +41,13 @@ typedef int (*validate_fn_t)(void*,void*,void*,int,char**);
+ *
+ * (Regular fields get cleared by config_reset(), but you might have fields
+ * in the object that do not correspond to configuration variables.  If those
+ * fields need to be cleared or freed, this is where to do it.)
```
What is the void * type here?

nmathewson replied:
Done in c281fa0da20ebb

teor2345 reviewed (changes requested):
I made some comments here, but GitHub seems to be having trouble finding some of them?
